File: sdk/typescript/src/loader.ts
/**
 * Data file loader with lazy caching and path resolution.
 */

import { readFileSync, existsSync, readdirSync, statSync } from "node:fs";
import { join, resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import type { DataFile } from "./types.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

let dataRoot: string | null = null;

function getDataRoot(): string {
  if (dataRoot !== null) return dataRoot;

  const envPath = process.env["HR_AGENT_DATA_PATH"];
  if (envPath) {
    dataRoot = resolve(envPath);
    return dataRoot;
  }

  // Relative to package root: sdk/typescript/src -> ../../data
  dataRoot = resolve(__dirname, "../../../data");
  return dataRoot;
}

function resolvePath(name: string): string {
  const root = getDataRoot();
  const fileName = name.endsWith(".json") ? name : `${name}.json`;
  const filePath = join(root, fileName);
  if (!existsSync(filePath)) {
    throw new Error(`Data file not found: ${filePath}`);
  }
  return filePath;
}

const cache = new Map<string, unknown>();

/**
 * Load and cache a data file by dotted name.
 *
 * @param name - Dotted path relative to data/, e.g. "offboarding/taxonomy"
 * @returns Parsed JSON data
 */
export function load<T = DataFile>(name: string): T {
  const cached = cache.get(name);
  if (cached !== undefined) return cached as T;

  const filePath = resolvePath(name);
  const content = readFileSync(filePath, "utf-8");
  const data = JSON.parse(content) as T;
  cache.set(name, data);
  return data;
}

/**
 * Load a data file as raw text.
 */
export function loadRaw(name: string): string {
  const filePath = resolvePath(name);
  return readFileSync(filePath, "utf-8");
}

/**
 * List available data files.
 */
export function listDataFiles(directory = ""): string[] {
  const root = directory ? join(getDataRoot(), directory) : getDataRoot();
  if (!existsSync(root)) return [];

  const files: string[] = [];

  function walk(dir: string): void {
    for (const entry of readdirSync(dir)) {
      const fullPath = join(dir, entry);
      const stat = statSync(fullPath);
      if (stat.isDirectory()) {
        walk(fullPath);
      } else if (entry.endsWith(".json") && !entry.startsWith("_")) {
        const relative = fullPath.replace(getDataRoot() + "/", "").replace(".json", "");
        files.push(relative);
      }
    }
  }

  walk(root);
  return files;
}

/**
 * Clear the in-memory cache. Useful for testing.
 */
export function clearCache(): void {
  cache.clear();
  dataRoot = null;
}

File: sdk/python/src/hr_agent_datagraph/loader.py
"""Data file loader with lazy caching and path resolution."""

from __future__ import annotations

import json
import os
from functools import lru_cache
from pathlib import Path
from typing import Any

_DATA_ROOT: Path | None = None


def _get_data_root() -> Path:
    """Resolve the data directory root.

    Resolution order:
    1. HR_AGENT_DATA_PATH environment variable
    2. <package_root>/../../data/ (relative to this file)
    """
    global _DATA_ROOT
    if _DATA_ROOT is not None:
        return _DATA_ROOT

    env_path = os.environ.get("HR_AGENT_DATA_PATH")
    if env_path:
        _DATA_ROOT = Path(env_path).resolve()
        return _DATA_ROOT

    _DATA_ROOT = (Path(__file__).resolve().parent.parent.parent.parent.parent / "data").resolve()
    return _DATA_ROOT


def _resolve_path(name: str) -> Path:
    """Resolve a dotted name to a data file path.

    Examples:
        "offboarding/taxonomy" -> data/offboarding/taxonomy.json
        "foundation/administrative-divisions/provinces" -> data/foundation/administrative-divisions/provinces.json
    """
    root = _get_data_root()
    if not name.endswith(".json"):
        name = f"{name}.json"
    path = root / name
    if not path.exists():
        raise FileNotFoundError(f"Data file not found: {path}")
    return path


@lru_cache(maxsize=64)
def load(name: str) -> dict[str, Any]:
    """Load and cache a data file by dotted name.

    Args:
        name: Dotted path relative to data/, e.g. "offboarding/taxonomy"

    Returns:
        Parsed JSON data as a dictionary.
    """
    path = _resolve_path(name)
    with open(path, encoding="utf-8") as f:
        return json.load(f)


def load_raw(name: str) -> str:
    """Load a data file as raw text.

    Args:
        name: Dotted path relative to data/

    Returns:
        Raw file content as string.
    """
    path = _resolve_path(name)
    return path.read_text(encoding="utf-8")


def list_data_files(directory: str = "") -> list[str]:
    """List available data files.

    Args:
        directory: Optional subdirectory to list, e.g. "offboarding"

    Returns:
        List of data file names (without .json extension).
    """
    root = _get_data_root() / directory if directory else _get_data_root()
    if not root.exists():
        return []
    return [
        str(p.relative_to(_get_data_root())).replace(".json", "")
        for p in root.rglob("*.json")
        if not p.name.startswith("_")
    ]
